src/nvim/buffer.c

```diff
@@ -2954,76 +3002,129 @@ build_stl_str_hl (
+        out_p = out_p - n + 1;
         /* Fill up space left over by half a double-wide char. */
         while (++l < item[groupitem[groupdepth]].minwid)
-          *p++ = fillchar;
+          *out_p++ = fillchar;
+        // }
 
         /* correct the start of the items for the truncation */
         for (l = groupitem[groupdepth] + 1; l < curitem; l++) {
+          // Shift everything back by the number of removed bytes
           item[l].start -= n;
+
+          // If the item was partially or completely truncated, set its
+          // start to the start of the group
           if (item[l].start < t)
             item[l].start = t;
         }
+      // If the group is shorter than the minimum width, add padding characters.
       } else if (abs(item[groupitem[groupdepth]].minwid) > l) {
-        /* fill */
-        n = item[groupitem[groupdepth]].minwid;
-        if (n < 0) {
-          /* fill by appending characters */
-          n = 0 - n;
-          while (l++ < n && p + 1 < out + outlen)
-            *p++ = fillchar;
+        long min_group_width = item[groupitem[groupdepth]].minwid;
+        // If the group is left-aligned, add characters to the right.
+        if (min_group_width < 0) {
+          min_group_width = 0 - min_group_width;
+          while (l++ < min_group_width && out_p < out_end_p)
+            *out_p++ = fillchar;
+        // If the group is right-aligned, shift everything to the right and
+        // prepend with filler characters.
         } else {
-          /* fill by inserting characters */
-          memmove(t + n - l, t, (size_t)(p - t));
-          l = n - l;
-          if (p + l >= out + outlen)
-            l = (long)((out + outlen) - p - 1);
-          p += l;
-          for (n = groupitem[groupdepth] + 1; n < curitem; n++)
+          // { Move the group to the right
+          memmove(t + min_group_width - l, t, (size_t)(out_p - t));
+          l = min_group_width - l;
+          if (out_p + l >= (out_end_p + 1)) {
+            l = (long)(out_end_p - out_p);
+          }
+          out_p += l;
+          // }
+
+          // Adjust item start positions
+          for (int n = groupitem[groupdepth] + 1; n < curitem; n++) {
             item[n].start += l;
-          for (; l > 0; l--)
+          }
+
+          // Prepend the fill characters
+          for (; l > 0; l--) {
             *t++ = fillchar;
+          }
         }
       }
       continue;
     }
-    minwid = 0;
-    maxwid = 9999;
-    zeropad = FALSE;
-    l = 1;
-    if (*s == '0') {
-      s++;
-      zeropad = TRUE;
+    int minwid = 0;
+    int maxwid = 9999;
+    bool left_align = false;
+
+    // Denotes that numbers should be left-padded with zeros
+    bool zeropad = (*fmt_p == '0');
+    if (zeropad) {
+      fmt_p++;
     }
-    if (*s == '-') {
-      s++;
-      l = -1;
+
+    // Denotes that the item should be left-aligned.
+    // This is tracked by using a negative length.
```

Comment:
is the "negative length" comment still true? Seems like it is tracked via `left_align` boolean now.